File: examples/websocket-simple.ts
/**
 * Simple example of using the PolymarketWebSocketClient
 * 
 * This demonstrates the easiest way to subscribe to market data
 * using just your private key.
 */

import { ClobClient } from "@polymarket/clob-client";
import { Wallet } from "@ethersproject/wallet";
import { PolymarketWebSocketClient } from "../src/sdk";

async function main() {
	// Initialize CLOB client with your private key
	const host = "https://clob.polymarket.com";
	const signer = new Wallet(process.env.POLYMARKET_KEY!);
	const clobClient = new ClobClient(host, 137, signer);

	// Create WebSocket client with asset IDs to subscribe to
	const ws = new PolymarketWebSocketClient(clobClient, {
		assetIds: [
			"60487116984468020978247225474488676749601001829886755968952521846780452448915",
		],
		autoReconnect: true,
		debug: true,
	});

	// Register event handlers
	ws.on({
		onConnect: () => {
			console.log("✅ Connected to Polymarket WebSocket");
		},

		onBook: (msg) => {
			console.log(
				`📚 Book Update - Market: ${msg.market.substring(0, 10)}...`,
			);
			console.log(`   Bids: ${msg.bids.length}, Asks: ${msg.asks.length}`);
			
			if (msg.bids.length > 0) {
				const bestBid = msg.bids[msg.bids.length - 1];
				if (bestBid) {
					console.log(`   Best Bid: ${bestBid.price} (${bestBid.size})`);
				}
			}
			if (msg.asks.length > 0) {
				const bestAsk = msg.asks[0];
				if (bestAsk) {
					console.log(`   Best Ask: ${bestAsk.price} (${bestAsk.size})`);
				}
			}
		},

		onPriceChange: (msg) => {
			console.log(
				`💹 Price Change - ${msg.price_changes.length} change(s)`,
			);
			for (const change of msg.price_changes) {
				console.log(
					`   ${change.side} @ ${change.price} (size: ${change.size})`,
				);
			}
		},

		onTickSizeChange: (msg) => {
			console.log(
				`📏 Tick Size Changed: ${msg.old_tick_size} → ${msg.new_tick_size}`,
			);
		},

		onLastTradePrice: (msg) => {
			console.log(
				`💰 Trade: ${msg.side} @ ${msg.price} (size: ${msg.size})`,
			);
		},

		onError: (error) => {
			console.error("❌ WebSocket Error:", error.message);
		},

		onDisconnect: (code, reason) => {
			console.log(`👋 Disconnected: ${code} - ${reason}`);
		},

		onReconnect: (attempt) => {
			console.log(`🔄 Reconnecting... (attempt ${attempt})`);
		},
	});

	// Connect to WebSocket
	await ws.connect();

	console.log("📡 Listening for market data... Press Ctrl+C to exit");

	// Handle graceful shutdown
	process.on("SIGINT", () => {
		console.log("\n🛑 Shutting down...");
		ws.disconnect();
		process.exit(0);
	});
}

main().catch((error) => {
	console.error("Fatal error:", error);
	process.exit(1);
});

File: go-client/examples/websocket/websocket_simple.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/HuakunShen/polymarket-kit/go-client/client"
	"github.com/HuakunShen/polymarket-kit/go-client/types"
	"github.com/joho/godotenv"
)

func main() {
	// Load environment variables
	if err := godotenv.Load(); err != nil {
		log.Fatal("Error loading .env file")
	}

	privateKey := os.Getenv("POLYMARKET_KEY")
	if privateKey == "" {
		log.Fatal("POLYMARKET_KEY environment variable is required")
	}

	// Create CLOB client
	config := &client.ClientConfig{
		Host:          "https://clob.polymarket.com",
		ChainID:       types.ChainPolygon,
		PrivateKey:    privateKey,
		UseServerTime: true,
		Timeout:       30 * time.Second,
	}

	clobClient, err := client.NewClobClient(config)
	if err != nil {
		log.Fatalf("Failed to create CLOB client: %v", err)
	}

	fmt.Println("✅ CLOB client created successfully")

	// Create WebSocket client
	wsClient := client.NewWebSocketClient(clobClient, &client.WebSocketClientOptions{
		AssetIDs: []string{
			"60487116984468020978247225474488676749601001829886755968952521846780452448915",
		},
		AutoReconnect:        true,
		ReconnectDelay:       5 * time.Second,
		MaxReconnectAttempts: 10,
		Debug:                true,
	})

	// Register event handlers
	wsClient.On(&client.WebSocketCallbacks{
		OnConnect: func() {
			fmt.Println("✅ Connected to Polymarket WebSocket")
		},

		OnBook: func(msg *types.BookMessage) {
			fmt.Printf("📚 Book Update - Market: %s...\n", msg.Market[:10])
			fmt.Printf("   Bids: %d, Asks: %d\n", len(msg.Bids), len(msg.Asks))

			if len(msg.Bids) > 0 {
				bestBid := msg.Bids[len(msg.Bids)-1]
				fmt.Printf("   Best Bid: %s (%s)\n", bestBid.Price, bestBid.Size)
			}
			if len(msg.Asks) > 0 {
				bestAsk := msg.Asks[0]
				fmt.Printf("   Best Ask: %s (%s)\n", bestAsk.Price, bestAsk.Size)
			}
		},

		OnPriceChange: func(msg *types.PriceChangeMessage) {
			fmt.Printf("💹 Price Change - %d change(s)\n", len(msg.PriceChanges))
			for _, change := range msg.PriceChanges {
				fmt.Printf("   %s @ %s (size: %s)\n", change.Side, change.Price, change.Size)
			}
		},

		OnTickSizeChange: func(msg *types.TickSizeChangeMessage) {
			fmt.Printf("📏 Tick Size Changed: %s → %s\n", msg.OldTickSize, msg.NewTickSize)
		},

		OnLastTradePrice: func(msg *types.LastTradePriceMessage) {
			fmt.Printf("💰 Trade: %s @ %s (size: %s)\n", msg.Side, msg.Price, msg.Size)
		},

		OnError: func(err error) {
			fmt.Printf("❌ WebSocket Error: %v\n", err)
		},

		OnDisconnect: func(code int, reason string) {
			fmt.Printf("👋 Disconnected: %d - %s\n", code, reason)
		},

		OnReconnect: func(attempt int) {
			fmt.Printf("🔄 Reconnecting... (attempt %d)\n", attempt)
		},
	})

	// Connect to WebSocket
	if err := wsClient.Connect(); err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}

	fmt.Println("📡 Listening for market data... Press Ctrl+C to exit")

	// Wait for interrupt signal
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)

	<-sigChan

	fmt.Println("\n🛑 Shutting down...")
	wsClient.Disconnect()
	fmt.Println("✅ Disconnected gracefully")
}
